packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,184 @@
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.Dedicated;
+    }
+
+    export function isNewTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.New;
+    }
+
+    export function isSharedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig &&
+            (options.taskConfig.presentation === undefined || options.taskConfig.presentation.panel === undefined || options.taskConfig.presentation.panel === PanelKind.Shared);
+    }
+}
+
+@injectable()
+export class TaskTerminalWidgetManager {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    @inject(TaskDefinitionRegistry)
+    protected readonly taskDefinitionRegistry: TaskDefinitionRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    @inject(TaskWatcher)
+    protected readonly taskWatcher: TaskWatcher;
+
+    @inject(TaskServer)
+    protected readonly taskServer: TaskServer;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @postConstruct()
+    protected init(): void {
+        this.taskWatcher.onTaskExit((event: TaskExitedEvent) => {
+            const finishedTaskId = event.taskId;
+            // find the terminal where the task ran, and mark it as "idle"
+            for (const widget of this.getTaskTerminalWidgets()) {
+                if (widget.taskId === finishedTaskId) {
+                    widget.busy = false;
+                    widget.scrollToBottom();
+                    widget.writeLine('\x1b[1m\n\rTerminal will be reused by tasks. \x1b[0m\n');
+                    break;
+                }
+            }
+        });
+
+        const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
+        this.terminalService.onDidCreateTerminal((widget: TerminalWidget) => {
+            const toDisposeCreateListener = widget.onTerminalDidConnectProcess(async (connectedWidget: TerminalWidget) => {
+                const tasksInfo = await this.taskServer.getTasks(context);
+                const taskInfo = tasksInfo.find(info => info.terminalId === connectedWidget.terminalId);
+                if (taskInfo) {
+                    const taskConfig = taskInfo.config;
+                    const dedicated = !!taskConfig.presentation && !!taskConfig.presentation.panel && taskConfig.presentation.panel === PanelKind.Dedicated;
+                    Object.assign(connectedWidget, {
+                        dedicated,
+                        taskId: taskInfo.taskId,
+                        taskConfig,
+                        busy: true
+                    });
+                }
+                const toDisposeCloseEvent = connectedWidget.onDidDispose(() => {
```

Comment:
It has to go to outer if to dispose didConnect listener even if a terminal never connected